backend/uclapi/oauth/views.py

```diff
@@ -0,0 +1,347 @@
+            "private_uclu": app.scope.private_uclu
+        },
+        "user": {
+            "full_name": user.full_name,
+            "cn": user.cn,
+            "email": user.email,
+            "department": user.department,
+            "upi": user.employee_id
+        },
+        "signed_data": response_data_signed
+    }
+
+    initial_data = json.dumps(page_data, cls=DjangoJSONEncoder)
+    return render(request, 'permissions.html',
+    {
+        'initial_data': initial_data
+    })
+
+
+@csrf_protect
+def userdeny(request):
+    signer = TimestampSigner()
+
+    try:
+        signed_data = request.POST.get("signed_app_data")
+        raw_data_str = signer.unsign(signed_data, max_age=300)
+    except:
+        return JsonResponse({
+            "ok": False,
+            "error": "The signed data received was invalid. Please try the login process again. If this issue persists, please contact support."
+        })
+    
+    try:
+        data = json.loads(raw_data_str)
+    except:
+        return JsonResponse({
+            "ok": False,
+            "error": "The JSON data was not in the expected format. Please contact support."
+        })
+
+    app = App.objects.get(client_id=data["client_id"])
+    state = data["state"]
+
+    redir = app.callback_url + "/denied?state=" + state
+    
+    return redirect(redir)
+    
+@csrf_protect
+def userallow(request):
+    signer = TimestampSigner()
+
+    try:
+        raw_data_str = signer.unsign(request.POST.get("signed_app_data"), max_age=300)
+    except:
+         return JsonResponse({
+            "ok": False,
+            "error": "The signed data received was invalid. Please try the login process again. If this issue persists, please contact support."
+        })
+    
+    try:
+        data = json.loads(raw_data_str)
+    except:
+        return JsonResponse({
+            "ok": False,
+            "error": "The JSON data was not in the expected format. Please contact support."
+        })
+
+    user = User.objects.get(employee_id=data["user_upi"])
+    app = App.objects.get(client_id=data["client_id"])
+    state = data["state"]
+
+    # Now we have the data we need to generate a random code and send this to the backend along
+    # with the state and request the app's client secret. If this matches we can generate OAuth
+    # keys and pass them to the backend
+
+    code = generate_random_verification_code()
+    verification_data = {
+        "verification_code": code,
+        "client_id": app.client_id,
+        "state": state
+    }
+
+    verification_data_str = json.dumps(verification_data, cls=DjangoJSONEncoder)
+    verification_data_str_enc = signer.sign(verification_data_str)
+
+    full_verification_data = {
+        "client_id": app.client_id,
+        "state": state,
+        "verification_data": verification_data_str_enc
+    }
+
+    try:
+        vr = requests.post(app.callback_url + "/verify", data=full_verification_data)
```

Comment:
okay. That's something to implement later I think :P Going to have to revamp the client library lol

Agreed though, as it needs to spec comply